File: mock-server.py
"""
mock-server.py — AeroGuard Python Mock Sensor API
==========================================================
Simulates the exact JSON shape a Raspberry Pi backend returns.
Zero dependencies beyond Python 3's standard library.

USAGE:
    python mock-server.py

Then in sensorManager.js set:
    const MOCK_MODE = false;
    const API_URL   = 'http://localhost:5000/api/sensor';
"""

import json
import math
import random
import time
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler, HTTPServer

PORT = 5000

# Shared random-walk state — gradual drift mirrors sensorManager.js mock
state = {"pm25": 18.0, "co2": 420.0, "voc": 120.0}


def next_reading():
    """Advance the random walk one step and return a sensor reading dict."""
    state["pm25"] = max(1.0,   min(300.0,  state["pm25"] + (random.random() - 0.48) * 3))
    state["co2"]  = max(350.0, min(3000.0, state["co2"]  + (random.random() - 0.48) * 15))
    state["voc"]  = max(0.0,   min(1000.0, state["voc"]  + (random.random() - 0.48) * 8))

    return {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "pm25":      round(state["pm25"], 1),
        "co2":       round(state["co2"]),
        "voc":       round(state["voc"]),
        # aqi is intentionally omitted — sensorManager.js computes it
    }


class SensorHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/api/sensor":
            body = json.dumps(next_reading()).encode()
            self.send_response(200)
            self.send_header("Content-Type",  "application/json")
            self.send_header("Content-Length", str(len(body)))
            # Allow cross-origin requests from the frontend
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(body)
        else:
            self.send_response(404)
            self.end_headers()

    # Suppress default per-request log noise (keep terminal clean during demo)
    def log_message(self, fmt, *args):
        pass


if __name__ == "__main__":
    server = HTTPServer(("0.0.0.0", PORT), SensorHandler)
    print(f"AeroGuard mock sensor API running at http://localhost:{PORT}/api/sensor")
    print("Press Ctrl+C to stop.")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nServer stopped.")

File: mock-server.js
// ============================================================
// mock-server.js — Minimal Express API Mock
//
// Simulates the exact JSON shape your Raspberry Pi backend
// must return. Use this to test the live-API path before
// your Pi is ready.
//
// SETUP:
//   npm install express
//   node mock-server.js
//
// Then in sensorManager.js:
//   const MOCK_MODE = false;
//   const API_URL   = 'http://localhost:5000/api/sensor';
//
// On your actual Pi, replicate this endpoint reading from your
// sensor library (e.g. pms5003, MH-Z19, SGP30) and return the
// same JSON shape. Zero other frontend changes needed.
// ============================================================

const express = require('express');
const app     = express();
const PORT    = 5000;

// Allow requests from the frontend (CORS)
app.use((req, res, next) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  next();
});

// Internal random-walk state (mirrors sensorManager.js mock for consistency)
let state = { pm25: 18.0, co2: 420, voc: 120 };

// GET /api/sensor — returns one reading in the canonical AeroGuard JSON shape
app.get('/api/sensor', (req, res) => {
  // Advance the random walk one step
  state.pm25 = Math.max(1,    Math.min(300,  state.pm25 + (Math.random() - 0.48) * 3));
  state.co2  = Math.max(350,  Math.min(3000, state.co2  + (Math.random() - 0.48) * 15));
  state.voc  = Math.max(0,    Math.min(1000, state.voc  + (Math.random() - 0.48) * 8));

  // This is the exact shape sensorManager.js#normalize() expects.
  // Your Pi script must return at minimum: { timestamp, pm25, co2, voc }
  res.json({
    timestamp: new Date().toISOString(),
    pm25:      +state.pm25.toFixed(1),
    co2:       Math.round(state.co2),
    voc:       Math.round(state.voc),
    // aqi is optional — sensorManager will compute it if absent
  });
});

app.listen(PORT, () => {
  console.log(`✅  AeroGuard mock sensor API running at http://localhost:${PORT}/api/sensor`);
  console.log(`    Press Ctrl+C to stop.`);
});
